### ISSUE TITLE
Question: GoInfo not working for HandleFunc() or nearly any other http function?

### ISSUE THREAD

**Issue by stefanos82** [fatih/vim-go#2105]
I use the latest HEAD vim-go with go version go1.11.2 linux/amd64 on a Debian testing 64-bit system and when I place my cursor under http.HandleFunc() or **nearly any** other http function, the `:GoInfo` does not return anything.

Is this the expected behavior?

**Comment by bhcleek** [comment]
It is not expected behavior. I just tried it and got the expected description.

Are you seeing problems in module or GOPATH mode? What value are you using for `g:go_info_mode`?

**Comment by stefanos82** [comment]
The behavior is the same in both module and GOPATH.

Now about the `g:go_info_mode` I haven't touched anything; I just installed it based on "Vim 8 packages" instructions.

**Comment by bhcleek** [comment]
Have you run `:GoInstallBinaries`?

**Comment by bhcleek** [comment]
Regardless of which version of vim-go you use, the binaries will need to be installed.

**Comment by stefanos82** [comment]
I think you did not understand me well, so let me rephrase myself to make things simpler and clearer if possible.

When you use the "Vim 8 package" method to setup `vim-go`, the very first thing you do after cloning the repo is to run `:GoInstallBinaries`.

So to keep it updated, you `git pull` the latest changes and then execute `:GoUpdateBinaries`.

As you imagine, I have done so already; that's when I have noticed the strange behavior.

I have never had this issue before, only after I fetched the updates when I first opened this ticket.

Does this sound clearer now @bhcleek ?

**Comment by bhcleek** [comment]
Thank you for the clarification.

There are a couple of things we can do to analyze this. First, echo the command info using `let g:go_echo_command_info=1`. With that set you should see either a `SUCCESS` or a `FAILED` message. If you don't see it for some reason, try running `:messages` to see if it's there. Then echo the command that was run by using `let g:go_debug=['shell-commands']`. That should output the command that vim-go runs as an array (again, you can see it with `:messages` if you miss it. You can then run the command (just join each element of the array); what is its output?

**Comment by bhcleek** [comment]
Does `:GoInfo` work for you for functions from other packages?

**Comment by stefanos82** [comment]
> Thank you for the clarification.
> 
> There are a couple of things we can do to analyze this. First, echo the command info using `let g:go_echo_command_info=1`. With that set you should see either a `SUCCESS` or a `FAILED` message. If you don't see it for some reason, try running `:messages` to see if it's there. Then echo the command that was run by using `let g:go_debug=['shell-commands']`. That should output the command that vim-go runs as an array (again, you can see it with `:messages` if you miss it. You can then run the command (just join each element of the array); what is its output?

This is what `:messages` says:
```
Messages maintainer: Bram Moolenaar <Bram@vim.org>
"main.go" 17L, 348C
vim-go: [gocode] dispatched
vim-go: [gocode] SUCCESS
```

and this is what I get with `:GoInfo` when `let g:go_debug=['shell-commands']` is being used:
```
vim-go: shell command: 'go' 'env' 'GOROOT'
vim-go: shell command: 'go' 'env' 'GOMOD'
vim-go: [gocode] dispatched
vim-go: job command: ['/home/stefanos/go/bin/gocode-gomod', '-sock', 'unix', '-f
', 'vim', '-builtin', '-fallback-to-source', '-cache', 'autocomplete', '/home/st
vim-go: func HandleFunc(pattern string, handler func(http.ResponseWriter, *http.
Request))
```

I have clearly see the function information, but as soon as I disable both `let g:go_echo_command_info=1` and `let g:go_debug=['shell-commands']`, I get nothing at all **for http.HandleFunc()** and pretty much, nearly all of `http` package function or method members.

For other functions and methods, for instance for `log.Fatal()`, it provides the information I'm looking for.

I think it's a well-hidden bug somewhere. It makes absolute no sense to work on some areas and not on some others.

**Comment by bhcleek** [comment]
That last line, `vim-go: func HandleFunc(pattern string, handler func(http.ResponseWriter, *http.
Request))` looks like the `:GoInfo` response you're looking for, but it _shouldn't_ be listed in `:messages`, because it is written with `echo` instead of `echom`. 🤔 Have you tried this with a minimal vimrc?

I suspect that if you run `/home/stefanos/go/bin/gocode-gomod -sock unix -f vim -builtin -fallback-to-source -cache autocomplete /home/stefanos/code/golang/webserver/main.go 81` you're going to see that the command itself is working correctly.

**Comment by stefanos82** [comment]
Nothing happens, which I presume it just starts the `gocode-gomod` to run as UNIX socket process and expects for input, even though I pass it the `.../webserver/main.go` along with line 81.

I'm sure though I'm doing it wrong.

Any other option?

I'm open for suggestions.

**Comment by bhcleek** [comment]
It's not waiting for input.

There are some more things we can try.

First, make sure to close all the gocodes that may be running:
```
$ gocode-gomod exit
$ gocode exit
$ gocode close #(just in case you have a  nsf/gocode running)
```

Then, start the server in a separate window with the `-debug` flag:
```
$ /home/stefanos/go/bin/gocode-gomod -sock unix -builtin -fallback-to-source -cache -debug
```
and then issue the client command:
```
$  /home/stefanos/go/bin/gocode-gomod -f vim autocomplete /home/stefanos/code/golang/webserver/main.go 81
```

**Comment by stefanos82** [comment]
> It's not waiting for input.
> 
> There are some more things we can try.
> 
> First, make sure to close all the gocodes that may be running:
> 
> ```
> $ gocode-gomod exit
> $ gocode exit
> $ gocode close #(just in case you have a  nsf/gocode running)
> ```

Done this step.

> 
> Then, start the server in a separate window with the `-debug` flag:
> 
> ```
> $ /home/stefanos/go/bin/gocode-gomod -sock unix -builtin -fallback-to-source -cache -debug
> ```

Done this as well.

> 
> and then issue the client command:
> 
> ```
> $  /home/stefanos/go/bin/gocode-gomod -f vim autocomplete /home/stefanos/code/golang/webserver/main.go 81
> ```

It hangs here...

![image](https://user-images.githubusercontent.com/1777466/50525062-f40e2d80-0ae1-11e9-82d1-de8aec1000fa.png)

Here is the background process:

![image](https://user-images.githubusercontent.com/1777466/50525019-b7dacd00-0ae1-11e9-8dfc-0b2f14f9eaf8.png)

**Comment by bhcleek** [comment]
bah, I gave bad instructions. Let me try that again.

For the server (added the `-s` flag):
```
$ /home/stefanos/go/bin/gocode-gomod -s -sock unix -builtin -fallback-to-source -cache -debug
```

For the client (added the `-sock unix` flag)
```
$  /home/stefanos/go/bin/gocode-gomod -sock unix -f vim autocomplete /home/stefanos/code/golang/webserver/main.go 81
```

**Comment by stefanos82** [comment]
Nothing happens...just blinking cursors to both terminals.

Am I supposed to expect something to happen?

**Comment by stefanos82** [comment]
> Sorry for the trouble: you need to add `-in /home/stefanos/code/golang/webserver/main.go` to the client command line since we're not sending the file's contents on stdin.

This did not do anything, like any difference at all.

> 
> FWIW, `:GoInfo` can use guru if you prefer by using `let g:go_info_mode='guru'`.

This however solved my problem!

Now, another interesting issue, which I don't know whether it relates somehow or not.

I ran `:map` to see which keys are available for use with Golang and could see that my already mapped `autocmd FileType go nmap <C-i> <Plug>(go-imports)` was also mapped with `<Tab>` key.

This mapping is taking place in `$HOME/.vim/ftplugin/gomappedkeys.vim`, along with two other mapped keys.

I said, that's cool; I could simply replace `(go-imports)` to `(go-info)` to make my life easier to get info any time I want to without having to run `:GoInfo` all the time by simply pressing `Ctrl-i`.

I did and tried to use the tab key to validate it behaves the same as `go-imports`; it did not.

It behaves the same as `go-info`.

Is this normal?

**Comment by bhcleek** [comment]
> This did not do anything, like any difference at all.

That's quite odd. You may want to raise an issue with github.com/mdempsky/gocode (for GOPATH mode) and github.com/stamblerre/gocode (for module mode).

Whether or not C-i behaves the same as Tab is outside the scope of vim-go. However, a quick search reveals https://github.com/neovim/neovim/issues/5916

If you're comfortable with the solution of using `guru` instead of `gocode`, please feel free to close this issue.

**Comment by stefanos82** [comment]
Thank you very much.

Cheers.
